proof-systems/src/darlin/dlog_acc_marlin/mod.rs

```diff
@@ -0,0 +1,936 @@
+                .map_err(Error::from_pc_err)?;
+        end_timer!(third_round_comm_time);
+
+        // again, absorb the prove oracles by the Fiat-Shamir rng
+        fs_rng.absorb(&to_bytes![third_comms].unwrap());
+
+        /* Preparations before entering the batch evaluation proof
+         */
+
+        let verifier_state = IOP::verifier_third_round(verifier_state, &mut fs_rng);
+
+        // Gather prover polynomials in one vector.
+        let mut polynomials: Vec<_> = index_pk
+            .0
+            .index
+            .iter()
+            .chain(prover_first_oracles.iter())
+            .chain(prover_second_oracles.iter())
+            .chain(prover_third_oracles.iter())
+            .collect();
+
+        // Compute bullet polynomial from the xi_s and we add it to the list of polynomials
+        let bullet_poly_coeffs = prev_acc.1[0].xi_s.compute_coeffs();
+        let bullet_poly = DensePolynomial::from_coefficients_vec(bullet_poly_coeffs);
+        let bullet_poly =
+            LabeledPolynomial::new("prev bullet poly".to_string(), bullet_poly, false);
+        polynomials.push(&bullet_poly);
+
+        // Gather commitments in one vector.
+        #[rustfmt::skip]
+        let commitments = vec![
+            first_comms.iter().map(|p| p.commitment().clone()).collect(),
+            second_comms.iter().map(|p| p.commitment().clone()).collect(),
+            third_comms.iter().map(|p| p.commitment().clone()).collect(),
+        ];
+        let mut labeled_comms: Vec<_> = index_pk
+            .0
+            .index_vk
+            .iter()
+            .cloned()
+            .zip(&IOP::<G1::ScalarField>::INDEXER_POLYNOMIALS)
+            .map(|(c, l)| LabeledCommitment::new(l.to_string(), c))
+            .chain(first_comms.iter().cloned())
+            .chain(second_comms.iter().cloned())
+            .chain(third_comms.iter().cloned())
+            .collect();
+
+        // Add g_fin to the labeled commitments
+        labeled_comms.push(LabeledCommitment::new(
+            "prev bullet poly".to_string(),
+            prev_acc.1[0].g_final.clone(),
+        ));
+
+        // Gather commitment randomness together.
+        let mut comm_rands: Vec<
+            LabeledRandomness<<PC<G1, D> as PolynomialCommitment<G1>>::Randomness>,
+        > = index_pk
+            .0
+            .index_comm_rands
+            .clone()
+            .into_iter()
+            .chain(first_comm_rands)
+            .chain(second_comm_rands)
+            .chain(third_comm_rands)
+            .collect();
+
+        // Add g_fin randomness (none because is public)
+        comm_rands.push(LabeledRandomness::new(
+            "prev bullet poly".to_string(),
+            <PC<G1, D> as PolynomialCommitment<G1>>::Randomness::zero(),
+        ));
+
+        // Set up the IOP verifier's query set.
+        let (mut query_set, verifier_state) = IOP::verifier_query_set(verifier_state)?;
```

Comment:
As the discussion we had some time ago: we can generalize everything for sure but it's not really worth it at the moment and will lead to other time spent in refactoring rather than implementing core functionalities.
So let's postpone this need of generalizing Marlin/IOP, perhaps I will open an issue about that.